HBase improvements

GerardDellemann opened the pull request:
HBase improvements - Added row-deletion functionality, also made HBaseDataContext implement UpdateableDataContext

Commit:
Notice with the builders that I tried to follow the setup used in other projects, but this often didn't work for HBase's column-families.
The FindQualifiersDriver works, but only from command line within a VM which runs Hadoop + HBase. It shows HBase working in combination with MapReduce. The class will be improved so it's method can be called by other classes.
All other classes have been tested by connecting it to our DataCleaner. I'll make more Unit Tests next.
hbase/pom.xml changes: With 'hbase-client' I needed to remove 2 exclusion. I've added 'hbase-server' dependency.

Commit:
Small improvements and fixed the existing Unit Test

arjansh committed:
Reverted changes to logic in core module and refactored HBaseColumn class so it now extends the AbstractColumn class (and implement the Column class).

GerardDellemann committed:
Merge pull request #1 from arjansh/feature/hbase-improvements-refactoring

Reverting some core changes and refactoring HBaseColumn class

Commit:
HBase improvements - Added row-deletion functionality, also made HBaseDataContext implement UpdateableDataContext

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseColumn.java:
```diff
@@ -0,0 +1,134 @@
+    private final ColumnType columnType;
+    private final int columnNumber;
+
+    public HBaseColumn(String columnFamily, Table table) {
```
In general try to make all parameters final, so it should look like this:

```
    public HBaseColumn(final String columnFamily, final Table table) {
```

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,71 @@
+            HBaseColumn[] outputColumns) {
+        super(updateCallback, schema, name);
+        if (!(schema instanceof MutableSchema)) {
+            throw new IllegalArgumentException("Not a valid schema: " + schema);
```
Maybe change "Not a valid schema: " to "Not a mutable schema: ".

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,71 @@
+    private Set<String> columnFamilies;
+
+    public HBaseCreateTableBuilder(HBaseUpdateCallback updateCallback, Schema schema, String name,
+            HBaseColumn[] outputColumns) {
```
Can you rename the `outputColumns` parameter to `columns`?

arjansh replied:
You only use the column families on the `HBaseColumn` for creating the table and the `HBaseColumn` objects themselves can't be constructed without an `HTable` object, so I think it's better to pass the column families as an argument here instead of `HBaseColumn` objects.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public Table execute() throws MetaModelException {
+        final MutableTable table = getTable();
```
There is no need for the `table` variable to be a `MutableTable`, it can also just be a `Table`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,71 @@
+        final SimpleTableDef emptyTableDef = new SimpleTableDef(table.getName(), columnFamilies.toArray(
+                new String[columnFamilies.size()]));
+
+        final HBaseUpdateCallback updateCallback = (HBaseUpdateCallback) getUpdateCallback();
```
There is no need to cast the result of the getUpdateCallback() method to `HBaseUpdateCallback`, it already is an `HBaseUpdateCallback` object because of the type parameter in this class's definition, which the getUpdateCallback() method uses for its return type.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,71 @@
+                    .getConfiguration()));
+            HbaseWriter.createTable(table.getName(), columnFamilies);
+        } catch (IOException e) {
+            e.printStackTrace();
```
This seems like a good place to throw a `MetaModelException`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseDataContext.java:
```diff
@@ -123,28 +125,28 @@ public Connection getConnection() {
 
     @Override
-    protected Schema getMainSchema() throws MetaModelException {
+    public Schema getMainSchema() throws MetaModelException {
```
Why make this method public? Why not leave it protected?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseDataContext.java:
```diff
@@ -95,7 +97,7 @@ private Connection createConnection(Configuration config) {
     }
 
-    private Configuration createConfig(HBaseConfiguration configuration) {
+    static protected Configuration createConfig(HBaseConfiguration configuration) {
```
Reorder the modifiers to comply with the Java Language Specification. It should be `protected static` instead of `static protected`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseDataContext.java:
```diff
@@ -123,28 +125,28 @@ public Connection getConnection() {
                     SimpleTableDef emptyTableDef = new SimpleTableDef(tables[i].getNameAsString(), new String[0]);
                     tableDefinitions[i] = emptyTableDef;
                 }
+            } catch (IllegalArgumentException | IOException e) {
```
Do you really want to catch an `IllegalArgumentException` here? I can't tell what throws it.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRow.java:
```diff
@@ -59,7 +59,6 @@ public Object getValue(int index) throws IndexOutOfBoundsException {
             }
             return rowKey;
         }
-
```
If you're only removing one empty line, I would just not bother doing that at all.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowDeletionBuilder.java:
```diff
@@ -0,0 +1,50 @@
+
+public class HBaseRowDeletionBuilder extends AbstractRowDeletionBuilder {
+
+    private HBaseUpdateCallback _updateCallback;
```
Please make the `_updateCallback` field final (and personally I would remove the `_` in front of it.)

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowDeletionBuilder.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public void execute() throws MetaModelException {
```
Remove the declaration of thrown exception 'org.apache.metamodel.MetaModelException' which is a runtime exception.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowDeletionBuilder.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public void execute() throws MetaModelException {
+        if (_key == null) {
+            throw new IllegalArgumentException("Key cannot be null");
```
`_key` is not an argument of this method so it doesn't seem logical to throw an `IllegalArgumentException` here, you can throw a `MetaModelException` instead.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+
+    private final HBaseColumn[] _outputColumns;
+
+    public HBaseRowInsertionBuilder(HBaseUpdateCallback updateCallback, HBaseTable table, HBaseColumn[] outputColumns) {
```
Here I would also rename the `outputColumns` variable to `columns`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+
+public class HBaseRowInsertionBuilder extends AbstractRowInsertionBuilder<HBaseUpdateCallback> {
+
+    private final HBaseColumn[] _outputColumns;
```
I would rename the `_outputColumns` field to `columns` (or `_columns`).

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    @Override
+    public void execute() throws MetaModelException {
```
Remove the declaration of thrown exception 'org.apache.metamodel.MetaModelException' which is a runtime exception.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Override
+    public void execute() throws MetaModelException {
+        checkForMatchingColumnFamilies(getTable(), _outputColumns);
```
Maybe move this check into the HBaseUpdateCallback#writeRow method and have that throw the `IllegalArgumentException` if the check fails.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    public HBaseColumn[] getOutputColumns() {
```
Do we really need this getter? And if we do, can't we call it `getColumns` instead?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+        return _outputColumns;
+    }
+
+    public void setOutputColumns(HBaseColumn[] outputColumns) {
```
Do we really need this setter? And if we do, can't we call it `setColumns` instead?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseTableDropBuilder.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Override
+    public void execute() throws MetaModelException {
```
Remove the declaration of thrown exception 'org.apache.metamodel.MetaModelException' which is a runtime exception.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+
+public class HBaseUpdateCallback extends AbstractUpdateCallback implements UpdateCallback {
+
+    private final HBaseConfiguration _configuration;
```
You can retrieve the `HBaseConfiguration` by invoking `getConfiguration` on the `HBaseDataContext`, so I see no need to store it in a field.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+
+    private final HBaseConfiguration _configuration;
+
+    private final HBaseDataContext _dataContext;
```
The `_dataContext` is already stored in the parent's class and accessible through the `AbstractUpdateCallback#getDataContext` method, so I would remove the field here. You can override the `getDataContext` method in this class with one that casts the returned `DataContext` object to the `HBaseDataContext` class.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    @Override
+    public TableCreationBuilder createTable(Schema schema, String name) throws IllegalArgumentException,
```
Remove the declarations of thrown exceptions 'java.lang.IllegalArgumentException' and 'java.lang.IllegalStateException' which are runtime exceptions.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    public HBaseCreateTableBuilder createTable(Schema schema, String name, HBaseColumn[] outputColumns)
+            throws IllegalArgumentException, IllegalStateException {
```
Remove the declarations of thrown exceptions 'java.lang.IllegalArgumentException' and 'java.lang.IllegalStateException' which are runtime exceptions.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public RowInsertionBuilder insertInto(Table table) throws IllegalArgumentException, IllegalStateException,
+            UnsupportedOperationException {
```
Remove the declarations of thrown exceptions 'java.lang.IllegalArgumentException', 'java.lang.IllegalStateException' and 'java.lang.UnsupportedOperationException' which are runtime exceptions.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+        throw new UnsupportedOperationException("Use insertInto(Table table, HBaseColumn[] outputColumns)");
+    }
+
+    public HBaseRowInsertionBuilder insertInto(Table table, HBaseColumn[] columns) throws IllegalArgumentException {
```
Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+
+    @Override
+    public RowDeletionBuilder deleteFrom(Table table) throws IllegalArgumentException, IllegalStateException,
+            UnsupportedOperationException {
```
Remove the declarations of thrown exceptions 'java.lang.IllegalArgumentException', 'java.lang.IllegalStateException' and 'java.lang.UnsupportedOperationException' which are runtime exceptions.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    public HBaseRowDeletionBuilder deleteFrom(HBaseUpdateCallback updateCallback, Table table)
+            throws IllegalArgumentException {
```
Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a runtime exception.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+        return new HBaseTableDropBuilder(table, this);
+    }
+
+    public void dropTableExecute(Table table) {
```
Can you remove this method and move it's logic into `HBaseTableDropBuilder#execute`?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+            MutableSchema schema = (MutableSchema) table.getSchema();
+            schema.removeTable(table);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Can you remove this and instead throw a `MetaModelException` which wraps the `IOException`?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+        throw new IllegalArgumentException("Not an HBase table: " + table);
+    }
+
+    protected synchronized void writeRow(HBaseTable hBaseTable, HBaseColumn[] outputColumns, Object[] values) {
```
Why is this method `protected`? To me `protected` indicates that it is a method which is typically overridden in a sub-class, but that's not the case here, as far as I can tell you only need it to be accessible from the `HBaseRowInsertionBuilder`. So than it could also be package protected instead, but I personally think it would even be better to remove this method and move its logic into the `HBaseRowInsertionBuilder#execute` method.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+        throw new IllegalArgumentException("Not an HBase table: " + table);
+    }
+
+    protected synchronized void deleteRow(HBaseTable hBaseTable, Object key) {
```
Same here as for the `writeRow` method, why is this method `protected`? To me `protected` indicates that it is a method which is typically overridden in a sub-class, but that's not the case here, as far as I can tell you only need it to be accessible from the `HBaseRowDeletionBuilder`. So than it could also be package protected instead, but I personally think it would even be better to remove this method and move its logic into the `HBaseRowDeletionBuilder#execute` method.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+            final HBaseWriter HbaseWriter = new HBaseWriter(HBaseDataContext.createConfig(_configuration));
+            HbaseWriter.writeRow(hBaseTable, outputColumns, values);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Can you remove this and instead throw a `MetaModelException` which wraps the `IOException`?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,137 @@
+            final HBaseWriter HbaseWriter = new HBaseWriter(HBaseDataContext.createConfig(_configuration));
+            HbaseWriter.deleteRow(hBaseTable, key);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Can you remove this and instead throw a `MetaModelException` which wraps the `IOException`?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseWriter.java:
```diff
@@ -0,0 +1,147 @@
+        try {
+            Table table = _connection.getTable(TableName.valueOf(hBaseTable.getName()));
+            try {
+                int indexOfIdColumn = 0;
```
This part where you determine the index of the id columns is pretty distracting and can be separated from the rest of the logic in this method, I think it would be a good idea to extract lines 55 through 66 into a new method called something like `getIndexOfIdColumn`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseWriter.java:
```diff
@@ -0,0 +1,147 @@
+
+    private boolean rowExists(Table table, Object key) throws IOException {
+        Get get = new Get(Bytes.toBytes(key.toString()));
+        return (table.get(get).isEmpty()) == true ? false : true;
```
`table.get(get).isEmpty()` already returns a boolean, so I think you can simplify this (and making it easier to understand) by changing it to 

```
return !table.get(get).isEmpty();
```

arjansh reviewed (changes requested):
Outside of the specific inline comments some more general comments (some of which are also mentioned within the specific inline comments)":
- In general try to make all parameters final.
- Try not to declare thrown runtime exceptions in method declarations.
- We may want to think through what kind of value to set for the ColumnType on the HBaseColumn a bit more (and in the same manner also what the getNativeType method returns).
- And we need unit tests for the added functionality.

arjansh committed:
Made getColumnsInternal method synchronized to prevent ConcurrentModificationExceptions when columns are retrieved from the Table in an parallel manner.

Commit:
HBase improvements: refactoring, new unittests

arjansh committed:
Refactored HBaseRowInsertionBuilder, so it uses the _columns of the AbstractRowBuilder instead of its own, to make sure the model stays consistent.

arjansh committed:
Removed table exists check. Let's assume the createTable message throws an exception when it fails to create the table.

Commit:
HBase Improvements - did some refactoring, added comments, added unittest

Commit:
HBase unittests improvements (fixed issues when the testfile wasn't setup)

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+
+    private final Connection _connection;
+
+    public HBaseClient(Connection connection) {
```
Can you make the `connection` parameter final?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @throws MetaModelException when a {@link IOException} is catched
+     */
+    // TODO: Use the ColumnTypes to determine the inserts. Now the toString() method is called on the object.
+    public void insertRow(String tableName, HBaseColumn[] columns, Object[] values, int indexOfIdColumn) {
```
Same here as above, can you make all parameters final?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @throws MetaModelException when no ID-column is found.
+     * @throws MetaModelException when a {@link IOException} is catched
+     */
+    // TODO: Use the ColumnTypes to determine the inserts. Now the toString() method is called on the object.
```
I don't think we can use the ColumnTypes here to determine the types for the inserts, because the types of the objects in the values array aren't determined by the ColumnType of their respective entries in the columns array. An object in the values array may be an Integer, in which case the ColumnType of its respective entries in the columns array will still be a Byte[], because we write everything to HBase as a Byte array.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+            // Add the other values to the put
+            for (int i = 0; i < columns.length; i++) {
+                if (i != indexOfIdColumn) {
+                    put.addColumn(Bytes.toBytes(columns[i].getColumnFamily()), Bytes.toBytes(columns[i].getQualifier()),
```
What do we do when an HBaseColumn doesn't have both a column family and qualifier set on it? Will the Put fail (partially?)? Or do we check that here and skip that column or thrown an exception?

GerardDellemann replied:
The columnFamily can't be null. This is forced by it's HBaseColumn's constructor. The class also doesn't have a set method on the variable.
In HBase it's possible to insert directly on the columnFamily without using the qualifier. I wanted to force this in the constuctor of HBaseColumn, but this couses with the HBaseDataContextTest#testCreateInsertQueryAndDrop test.
Then the assertEquals would need to be:
assertEquals("[" + HBaseDataContext.FIELD_ID + ":, " + CF_BAR + ":, " + CF_FOO + ":]",  ...
I would like to discus this.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+            for (int i = 0; i < columns.length; i++) {
+                if (i != indexOfIdColumn) {
+                    put.addColumn(Bytes.toBytes(columns[i].getColumnFamily()), Bytes.toBytes(columns[i].getQualifier()),
+                            Bytes.toBytes(values[i].toString()));
```
For now we can only assume each value object's toString() method returns a good representation of that object, but in case we're writing Lists or Maps, this may not be adequate.

GerardDellemann replied:
For Maps, see the result of the test. Let's discus if and how we need improvements

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @throws IllegalArgumentException when any parameter is null
+     * @throws MetaModelException when a {@link IOException} is catched
+     */
+    public void deleteRow(String tableName, Object rowKey) {
```
Same here as above, can you make all parameters final?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+        if (tableName == null || rowKey == null) {
+            throw new IllegalArgumentException("Can't delete a row without having tableName or rowKey");
+        }
+        try (final Table table = _connection.getTable(TableName.valueOf(tableName));) {
```
Can you remove the ugly `;`?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+            throw new IllegalArgumentException("Can't delete a row without having tableName or rowKey");
+        }
+        try (final Table table = _connection.getTable(TableName.valueOf(tableName));) {
+            if (rowExists(table, rowKey) == true) {
```
Isn't it redundant to compare with `true` here? Also, is it really necessary to check if it exists? What happens if it doesn't exist and the delete is executed anyway? If the delete can be executed without problems without checking if the row exists, does that make a significant difference in performance?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @param tableName
+     * @param rowKey
+     * @throws IllegalArgumentException when any parameter is null
+     * @throws MetaModelException when a {@link IOException} is catched
```
Past tense of catch is `caught`, not `catched`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @param values
+     * @throws IllegalArgumentException when any parameter is null or the indexOfIdColumn is impossible
+     * @throws MetaModelException when no ID-column is found.
+     * @throws MetaModelException when a {@link IOException} is catched
```
Past tense of catch is `caught`, not `catched`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @return boolean
+     * @throws IOException
+     */
+    private boolean rowExists(Table table, Object rowKey) throws IOException {
```
Same here as above, can you make all parameters final?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @throws IllegalArgumentException when any parameter is null
+     * @throws MetaModelException when a {@link IOException} is catched
+     */
+    public void createTable(String tableName, Set<String> columnFamilies) {
```
Same here as above, can you make all parameters final?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @param tableName
+     * @param columnFamilies
+     * @throws IllegalArgumentException when any parameter is null
+     * @throws MetaModelException when a {@link IOException} is catched
```
Past tense of catch is `caught`, not `catched`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * Disable and drop a table from a HBase datastore
+     * @param tableName
+     * @throws IllegalArgumentException when tableName is null
+     * @throws MetaModelException when a {@link IOException} is catched
```
Past tense of catch is `caught`, not `catched`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+     * @throws IllegalArgumentException when tableName is null
+     * @throws MetaModelException when a {@link IOException} is catched
+     */
+    public void dropTable(String tableName) {
```
Same here as above, can you make all parameters final?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class can perform client-operations on a HBase datastore
```
I would not refer to HBase as a datastore in this context, but as a database.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    /**
+     * Disable and drop a table from a HBase datastore
```
I would not refer to HBase as a datastore in this context, but as a database.

GerardDellemann replied:
Let's discus this.

Commit:
- Where missing added final qualifiers to method parameters.
- Moved static methods around to more logic locations. From my point of view the HBaseColumn class is not a Helper class to provide those.
- Changed the signature of the HBaseRowInsertionBuilder constructor, so it take a List of HBaseColumn object as a parameter, this makes it possible to remove a few of the Helper methods which cast List<HBaseColumn> to List<Column> and vice versa.
- Refactored the logic which looks up the id column in an array or list.
- Removed the "public" getHBaseColumnsInternal() method from HBaseTable, which essentially was only invoked in combination with a static getColumnFamilies method. Instead I added a getColumnFamilies() method to the HBaseTable class which return the column families for that HBase table.

GerardDellemann committed:
Merge pull request #3 from GerardDellemann/feature/hbase-improvements-refactoring

Refactoring code and implementing feedback

Commit:
HBase improvements (small refactorings, add comments)

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,115 @@
+        if (columnFamilies == null || columnFamilies.size() == 0) {
+            throw new MetaModelException("Creating a table without columnFamilies");
+        }
+        boolean idColumnFound = false;
```
It is not necessary to check if there is an `_id` column family, because this will be added implicitly when the table is created.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,115 @@
+     * Set the columnFamilies
+     * @param columnFamilies
+     */
+    public void setColumnFamilies(Set<String> columnFamilies) {
```
Why does this setter exist? It is only used in the unit tests which test it.

GerardDellemann replied:
The call hierachy doesn't suggest this, but HBaseUpdateCallback has 2 createTable(...) methods. The first one comes from the of inheritance "AbstractUpdateCallback". If we want to offering this (abstract) method, then we also need to set the columnFamilies before execution.
I've added JavaDoc to this this method to clearify this.

arjansh replied:
Well, in that case just make the `columnFamilies` parameter final.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseDataContext.java:
```diff
@@ -54,7 +56,7 @@
-public class HBaseDataContext extends QueryPostprocessDataContext {
+public class HBaseDataContext extends QueryPostprocessDataContext implements UpdateableDataContext {
 
     private static final Logger logger = LoggerFactory.getLogger(HBaseDataContext.class);
```
The value of the field `logger` is no longer used, so please remove the field too.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,142 @@
+            final List<HBaseColumn> columns) {
+        super(updateCallback, table, columns.stream().map(column -> (Column) column).collect(Collectors.toList()));
+        if (columns.isEmpty()) { // TODO: Columns null will already result in a NullPointer at the super. Should the
+                                 // super get a extra check?
```
I would not add such a check now. Maybe even leave out this one, because if `columns` is empty, you will get a MetaModel exception anyway because no _`id` column will be found.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,142 @@
+        }
+
+        checkTable(updateCallback, table);
+        checkForNotMatchingColumnFamilies(table, getColumnFamilies(columns));
```
I'm not sure what the extra value of calling `checkForNotMatchingColumnFamilies` is here? The method is already called from within the `checkTable` method, does it check something different here?

GerardDellemann replied:
It's possible to provide different columns from the table. Just to be sure this check is performed. I've added comments to explain this check.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,142 @@
+     * @param columnNamesOfCheckedTable
+     * @throws MetaModelException If a column doesn't exist
+     */
+    public void checkForNotMatchingColumnFamilies(final HBaseTable table, final Set<String> columnNamesOfCheckedTable) {
```
I really don't like the method name `checkForNotMatchingColumnFamilies`, maybe just call it `checkColumnFamilies` instead?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowInsertionBuilder.java:
```diff
@@ -0,0 +1,142 @@
+
+        for (String columnNameOfCheckedTable : columnNamesOfCheckedTable) {
+            boolean matchingColumnFound = false;
+            Iterator<String> iterator = columnFamilyNamesOfExistingTable.iterator();
```
Can you use a less abstract name then `iterator` for the variable, maybe `columnFamilyNames` or `columnFamilies`?

Commit:
HBase improvements (when creating a HBase table, the ID-column is not a required parameter, because that's always created).

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseTable.java:
```diff
@@ -43,85 +44,121 @@
         super(tableDef.getName(), TableType.TABLE, schema);
         _dataContext = dataContext;
         _defaultRowKeyColumnType = defaultRowKeyColumnType;
+        addColumns(tableDef, defaultRowKeyColumnType);
```
Why add `defaultRowKeyColumnType` as a parameter to the `addColumns` method if you assign it to the `_defaultRowKeyColumnType` field in the line above. I would remove the  `defaultRowKeyColumnType` as a parameter from the `addColumns` method and just use the value of the `_defaultRowKeyColumnType` field instead.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseTable.java:
```diff
@@ -43,85 +44,121 @@
+                // Add the ID-Column (with columnNumber = 1)
+                addColumn(HBaseDataContext.FIELD_ID, _defaultRowKeyColumnType, 1);
 
                 // What about timestamp?
```
Yes, what about timestamp? Do we still need this comment, or do we have an answer to it?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseTableDropBuilder.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public void execute() {
+        // Remove from the datastore
```
You are removing it from the actual HBase database, why call it datastore here?

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * @throws IllegalArgumentException when table isn't a {@link HBaseTable}
```
This method doesn't thrown an `IllegalArgumentException`, it thows an UnsupportedOperationException`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseUpdateCallback.java:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    public HBaseClient getHBaseClient() {
```
I think access to this method should be `package private`, because we don't want to expose the HBaseClient as part of the MetaModel api.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,169 @@
+/**
+ * This class can perform client-operations on a HBase datastore
+ */
+public final class HBaseClient {
```
I think this class should be `package private`,  because we don't want to expose it as part of the MetaModel api.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowDeletionBuilder.java:
```diff
@@ -0,0 +1,61 @@
+     * @param table
+     * @throws IllegalArgumentException when the hBaseWriter is null
+     */
+    public HBaseRowDeletionBuilder(final HBaseClient hBaseWriter, final Table table) {
```
All similar builders, like the `HBaseCreateTableBuilder`, `HBaseTableDropBuilder` and `HBaseRowInsertionBuilder` have the `HBaseUpdateCallback` passed in in their constructor, not the `HBaseClient`, is there a good reason to do it differently here?

arjansh reviewed (changes requested):
Review of the implementation code. Review of test code will follow.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,169 @@
+                if (rowExists(table, rowKeyAsByteArray) == true) {
+                    table.delete(new Delete(rowKeyAsByteArray));
+                } else {
+                    logger.warn("Rowkey with value " + rowKey.toString() + " doesn't exist in the table");
```
For performance reasons it's better to use a Marker object when inserting variables into log messages, e.g.:
```
logger.warn("Rowkey with value {} doesn't exist in the table", rowKey.toString());
```

GerardDellemann replied:
Can you give me something to read on this (why it's needed and how to implement it)?

arjansh replied:
See https://rules.sonarsource.com/java/RSPEC-2629.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseColumn.java:
```diff
@@ -0,0 +1,123 @@
+import org.apache.metamodel.schema.SuperColumnType;
+import org.apache.metamodel.schema.Table;
+
+public final class HBaseColumn extends MutableColumn {
```
Extend a class that overrides equals and add fields without overriding equals in the subclass, and you run the risk of non-equivalent instances of your subclass being seen as equal, because only the superclass fields will be considered in the equality test.

GerardDellemann replied:
TODO

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseColumn.java:
```diff
@@ -0,0 +1,123 @@
+public final class HBaseColumn extends MutableColumn {
+    public final static ColumnType DEFAULT_COLUMN_TYPE_FOR_ID_COLUMN = new ColumnTypeImpl("BYTE[]",
+            SuperColumnType.LITERAL_TYPE);
+    public final static ColumnType DEFAULT_COLUMN_TYPE_FOR_COLUMN_FAMILIES = ColumnType.LIST;
```
Reorder the modifiers to comply with the Java Language Specification, so it should be `public static final`.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseCreateTableBuilder.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Override
+    public Table execute() {
+        if (_columnFamilies == null || _columnFamilies.size() == 0) {
```
Use isEmpty() to check whether the collection is empty or not. Using `Collection.size()` to test for emptiness works, but using `Collection.isEmpty()` makes the code more readable and can be more performant.

arjansh commented on hbase/src/main/java/org/apache/metamodel/hbase/HBaseRowDeletionBuilder.java:
```diff
@@ -0,0 +1,61 @@
+        _hBaseClient.deleteRow(getTable().getName(), _key);
+    }
+
+    public void setKey(Object _key) {
```
Rename `_key` parameter to `key` and make it final.

Commit:
Refactored test so they no longer extends TestCase (and as such are no longer JUnit 3, but JUnit 4 tests) and they use "assumeTrue" to check if they should run or not. This reduces a lot of boiler plate code, because now we don't need to call the "isConfigured" method in each test method.

Also added some extra null check in the code which cleans up after test, because some nullpointer exceptions were thrown when the tests were skipped.

GerardDellemann committed:
Merge pull request #4 from GerardDellemann/feature/hbase-improvements-assume-in-tests

Refactored to remove boiler plate isConfigured checks

Commit:
HBase improvements - Refactoring (mostly small improvements, the getHBaseClient method has been moved to HBaseDataContent)

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/CreateTableTest.java:
```diff
@@ -0,0 +1,209 @@
+        final ImmutableSchema immutableSchema = new ImmutableSchema(getSchema());
+        try {
+            getUpdateCallback().createTable(immutableSchema, TABLE_NAME).execute();
+            fail("Should get an exception that the schema isn't mutable");
```
There is a different manner to test for exceptions in JUnit4 using rules. 

You can add this `rule` as a field member to the class:
```
    @Rule
    public ExpectedException exception = ExpectedException.none();
```

Then here in this method, before executing the createTable, you then add:

```
        exception.expect(IllegalArgumentException.class);
        exception.expectMessage("Not a mutable schema: " + immutableSchema);
```

For more info on this go to https://github.com/junit-team/junit4/wiki/exception-testing (and http://baddotrobot.com/blog/2012/03/27/expecting-exception-with-junit-rule/index.html).

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/CreateTableTest.java:
```diff
@@ -0,0 +1,209 @@
+            hBaseCreateTableBuilder.setColumnFamilies(columnFamilies);
+            hBaseCreateTableBuilder.execute();
+            checkSuccesfullyInsertedTable();
+        } catch (Exception e) {
```
Why do you catch ` Exception` here? None of the invoked methods throws `Exception`, there is only one which throws `IOException`,  but that doesn't need to be caught, because this test method already throws `IOException`, so I would remove the `try ... catch` construct here completely.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/CreateTableTest.java:
```diff
@@ -0,0 +1,209 @@
+            hBaseCreateTableBuilder.setColumnFamilies(columnFamilies);
+            hBaseCreateTableBuilder.execute();
+            checkSuccesfullyInsertedTable();
+        } catch (Exception e) {
```
Same like above, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/CreateTableTest.java:
```diff
@@ -0,0 +1,209 @@
+        try {
+            getUpdateCallback().createTable(getSchema(), TABLE_NAME, columnFamilies).execute();
+            checkSuccesfullyInsertedTable();
+        } catch (Exception e) {
```
Here too, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/DeleteRowTest.java:
```diff
@@ -0,0 +1,181 @@
+            rowDeletionBuilder.setKey(RK_1);
+            rowDeletionBuilder.execute();
+            checkRows(false, false);
+        } catch (Exception e) {
```
Same like earlier, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/DeleteRowTest.java:
```diff
@@ -0,0 +1,181 @@
+            rowDeletionBuilder.setKey("");
+            rowDeletionBuilder.execute();
+            checkRows(false, false);
+        } catch (Exception e) {
```
Same like earlier, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/DeleteRowTest.java:
```diff
@@ -0,0 +1,181 @@
+            rowDeletionBuilder.setKey(RK_1);
+            rowDeletionBuilder.execute();
+            checkRows(false, false);
+        } catch (Exception e) {
```
Same like earlier, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/DropTableTest.java:
```diff
@@ -0,0 +1,84 @@
+                try (final Admin admin = getDataContext().getAdmin()) {
+                    assertFalse(admin.tableExists(TableName.valueOf(TABLE_NAME)));
+                }
+            } catch (Exception e) {
```
Same like earlier, don't catch the exception.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/DropTableTest.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    @Test
+    public void testDropTableSupported() {
+            assertTrue(getUpdateCallback().isDropTableSupported());
```
Indentation is off, please reformat file.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/HBaseTestCase.java:
```diff
@@ -43,14 +78,28 @@ protected void setUp() throws Exception {
+    }
+
+    /**
+     * Get's the test configuration file. An example file can be found at the root folder of this project.
```
It's "Gets", not "Get's".

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/HBaseUpdateCallbackTest.java:
```diff
@@ -0,0 +1,283 @@
+                // Check in the datastore
+                try (final Admin admin = getDataContext().getAdmin()) {
+                    assertFalse(admin.tableExists(TableName.valueOf(TABLE_NAME)));
+                } catch (IOException e) {
```
Don't catch the `IOException` here (and don't "fail" the test in case the `IOException` occurs), throw it instead. If we throw it, it will result in an "error" in the test when it occurs, which it what it is, if we catch it and "fail" the test, it will show up as a failing test, which it isn't because it didn't actually test our test case.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/HBaseUpdateCallbackTest.java:
```diff
@@ -0,0 +1,283 @@
+        // Check in the datastore
+        try (final Admin admin = getDataContext().getAdmin()) {
+            assertTrue(admin.tableExists(TableName.valueOf(TABLE_NAME)));
+        } catch (IOException e) {
```
Same here as above, don't catch the IOException here (and don't "fail" the test in case the IOException occurs).

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/HBaseUpdateCallbackTest.java:
```diff
@@ -0,0 +1,283 @@
+     * @return created {@link HBaseTable}
+     */
+    protected HBaseTable createHBaseTable(final String tableName, final String idColumn, final String columnFamily1,
+            final String columnFamily2, final String columnFamily3) {
```
I don't like this, can't you use one columnFamily parameter instead which uses the Varargs feature, so `final String... columnFamilies`?

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/HBaseUpdateCallbackTest.java:
```diff
@@ -0,0 +1,283 @@
+        }
+    }
+
+    protected Collection<Object> getToLittleValues(final LinkedHashMap<HBaseColumn, Object> row) {
```
Shouldn't this method be named `getTooLittleValues` instead?

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/HBaseUpdateCallbackTest.java:
```diff
@@ -0,0 +1,283 @@
+     * @param className
+     * @param methodName
+     */
+    protected void warnAboutANotExecutedTest(String className, String methodName) {
```
This method can be removed because it is no longer used.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/InsertRowTest.java:
```diff
@@ -0,0 +1,390 @@
+                    CF_BAR, false);
+            final List<HBaseColumn> columns = getHBaseColumnsFromRow(row);
+            getUpdateCallback().insertInto(existingTable, columns);
+        } catch (Exception e) {
```
Same like earlier, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/InsertRowTest.java:
```diff
@@ -0,0 +1,390 @@
+            setValuesInInsertionBuilder(row, rowInsertionBuilder);
+            rowInsertionBuilder.execute();
+            checkRows(true, true);
+        } catch (Exception e) {
```
Same like above, don't catch the exception but have the method throw `IOException` instead.

arjansh commented on hbase/src/test/java/org/apache/metamodel/hbase/InsertRowTest.java:
```diff
@@ -0,0 +1,390 @@
+            setValuesInInsertionBuilder(row, rowInsertionBuilder);
+            rowInsertionBuilder.execute();
+            checkRows(true, false);
+        } catch (Exception e) {
```
Same like above, don't catch the exception but have the method throw `IOException` instead.

Commit:
HBaseClient.java:
- Remove overabundant `;` on line 111 ("try (final Table table = _connection.getTable(TableName.valueOf(tableName));) {").
- Remove overabundant " == true" on line 112 ("if (rowExists(table, rowKeyAsByteArray) == true) {").
- On line 115 use a Marker object when inserting variables into log messages.
- Add final modifier on line 144 ("public void createTable(String tableName, Set<String> columnFamilies) {").

HBaseCreateTableBuilder.java:
- On line 81 make the parameters of the setColumnFamilies method final.
HBaseRow.java:
- Revert remmoval of empty line (62).
HBaseRowDeletionBuilder.java:
- The _dataContext field member has to be made final
- On line 58 make the key par
